[CL-1839] Merge FranceConnect accounts based on first and last names

alexander-cit opened the pull request:
The docs should be updated https://www.notion.so/citizenlab/Authentication-Verification-8578ab8c280e41dcb877c1b8b5d924b5


I still need to test all possible scenarios, but I think in general it should work.

alexander-cit committed:
[CL-1839] Merge FranceConnect accounts based on first and last names

alexander-cit committed:
[CL-1839] Don't reopen signUpIn modal when it's closed after error

alexander-cit commented on back/app/controllers/omniauth_callback_controller.rb:
```diff
@@ -19,6 +19,25 @@ def create
     end
   end
 
+  def failure
```
`failure` and `logout` look like added just because I've moved `auth_callback` under the private section (as it's not an action).

nTraum replied:
Thanks for doing this :+1:

brentguf commented on front/app/components/SignUpIn/SignUp/messages.ts:
```diff
@@ -128,6 +128,11 @@ export default defineMessages({
+  franceConnectMergingFailed: {
+    id: 'app.containers.SignUp.franceConnectMergingFailed',
+    defaultMessage:
+      'An account already exists with this email address.{br}{br}You cannot access the platform using FranceConnect as the personal details do not match. To log in using France Connect, you will have to first change your first name or last name on this platform to match your official details.{br}{br}You can log in as you normally do below.',
```
"To log in using France Connect" => "FranceConnect"

alexander-cit replied:
Hmm We have a space in `What is France Connect?` too. Let me ask Ben if we should change both.

brentguf replied:
It is FranceConnect, one word: https://franceconnect.gouv.fr/.

brentguf reviewed (commented):
A small one I saw

alexander-cit committed:
[CL-1839] Change "France Connect" to "FranceConnect" (Crowdin is updated too)

alexander-cit committed:
[CL-1839] Always propose user to sign in in case of FC error even if they tried to sign up

Commit:
Translations updated by CI (extract-intl)

alexander-cit committed:
[CL-1839] Always redirect to main page in case of error

`sso_pathname: '/'` - when sso_pathname is `/en/sign-in`, it's not
redirected to /en/sign-in and the error message is not shown
On the FE, this hack can be tested accessing this URL
http://localhost:3000/authentication-error?sso_response=true&sso_flow=signin&sso_pathname=%2F&error_code=france_connect_merging_failed
Note, that the modal is not shown with this URL
http://localhost:3000/authentication-error?sso_response=true&sso_flow=signin&sso_pathname=%2Fen%2Fsign-in&error_code=france_connect_merging_failed

Probaby, it would be possible to fix it on the FE, but it seems to be much more complicated.

alexander-cit committed:
[CL-1839] Refactor some BE methods

alexander-cit committed:
[CL-1839] Make error message consistent with other strings

alexander-cit commented:
@kogre could you take a look at the BE part?

lilkraftwerk commented on front/app/containers/App/index.tsx:
```diff
@@ -172,7 +172,7 @@ class App extends PureComponent<Props, State> {
       mobileNavbarRef: null,
       locale: null,
-      invitationDeclined: false,
+      signUpInModalClosed: false,
```
I added `signUpInModalOpened: boolean;` in my earlier PR, it's a few lines above this one. I think it does the same thing so we can probably just keep one of them. Also applies to the setState function passed down into the modal I believe

alexander-cit replied:
But I need to pass `false` to `openSignUpInModalIfNecessary` when the modal was closed (so, we don't need to open it again). And your hook runs when the `SignUpInModal` is rendered.

Could you maybe show how it can be implemented to make it clearer for me?

lilkraftwerk replied:
Yeah let's pair on it tomorrow, maybe I misunderstood 👍

lilkraftwerk reviewed (approved):
Frontend looks good. BE too but I didn't review that too closely. Only one comment, there's some duplicated logic between my earlier PR and this one, I think in one or two places you can just hook into the state I introduced and it should be the same thing

alexander-cit committed:
[CL-1839] Prevent showing modal with error after successfull form submit

alexander-cit committed:
[CL-1839] Add comments about SignUpIn modal state that manages callbacks

alexander-cit committed:
[CL-1839] Rename some methods to what they actually do

Identity cannot be created in this case because user is blank but
required by `belongs_to`

kogre commented on back/app/controllers/omniauth_callback_controller.rb:
```diff
@@ -27,7 +46,24 @@ def auth_callback(verify:, authver_method:)
+
+    if @user.nil?
+      @user = User.find_by_cimail(user_attrs.fetch(:email))
+      if @user && !authver_method.can_merge?(@user, user_attrs)
```
Good that you're delegating this to the actual method :+1: 

I was thinking that we could generalize this even further and also let the authver method decide how an existing user is matched. E.g. the current hard coded `find_by_cimail` could become something like `authver_method.find_matching_user(user_attrs)`. It would let us achieve the same thing as `can_merge?` (when nothing is returned, there is no match and we error the same way you do now) but would give more flexibility to the method to match in different ways, which would help us with methods that don't support email, for example (we've had requests for this).

Just sharing this as an aside, if this is tested and properly validated and there are other concerns I'm not thinkin about, then I'm fine with sticking to this approach.

alexander-cit replied:
Indeed, it would be simpler, but it seems we could not handle all the cases with `find_matching_user`.

E.g., if `find_matching_user` returns `nil`, what does it mean? That we cannot merge and we should show the message about it? Or that we should create a new user with this email?

Currently, if `find_by_cimail` returns `nil`, we create a new user.

```
   find_by_cimail
  /               \
nil                existing user
 |                        |
(1)
we create             can_merge?
a new user           /         \
                    (2)       (3)
             yes, we merge       no, we show error message about names


    find_matching_user
  /               \
 (1)
nil                existing user
 |                        |
                         (2)
we create             we merge
a new user
```

So, there are three cases, and it seems `find_matching_user` can handle only two. But maybe I'm missing something.

kogre replied:
Yes, you're right. We could still do it through return values from `find_matching_user`, letting the method decide between these 3 cases (and maybe there'll be more in the future even). But I'd consider that out of scope for now.

kogre commented on back/app/controllers/omniauth_callback_controller.rb:
```diff
@@ -27,7 +46,24 @@ def auth_callback(verify:, authver_method:)
+    if @user.nil?
+      @user = User.find_by_cimail(user_attrs.fetch(:email))
+      if @user && !authver_method.can_merge?(@user, user_attrs)
+        # `sso_flow: 'signin'` - even if user signs up, we propose to sign in due to the content of the error message
+        #
+        # `sso_pathname: '/'` - when sso_pathname is `/en/sign-in`, it's not redirected to /en/sign-in and the error message is not shown
+        # On the FE, this hack can be tested accessing this URL
+        # http://localhost:3000/authentication-error?sso_response=true&sso_flow=signin&sso_pathname=%2F&error_code=franceconnect_merging_failed
+        # Note, that the modal is not shown with this URL
+        # http://localhost:3000/authentication-error?sso_response=true&sso_flow=signin&sso_pathname=%2Fen%2Fsign-in&error_code=franceconnect_merging_failed
+        #
+        # Probaby, it would be possible to fix both issues on the FE, but it seems to be much more complicated.
+        failure_redirect(error_code: authver_method.merging_error_code, sso_flow: 'signin', sso_pathname: '/')
+        return
```
Thanks for the comment explaining it, but it does sound a bit messy indeed, hope we can solve this at some point.

kogre commented on back/engines/commercial/id_franceconnect/spec/lib/id_franceconnect/franceconnect_omniauth_spec.rb:
```diff
@@ -0,0 +1,51 @@
+
+describe IdFranceconnect::FranceconnectOmniauth do
+  describe '#can_merge?' do
+    using RSpec::Parameterized::TableSyntax
```
Didn't know this way of providing examples, pretty nice, will remember it :)

kogre reviewed (approved):
Only reviewed the BE part. One larger comment, but not blocking to me. :+1:

alexander-cit committed:
[CL-1839] Always merge in case of verification (not authorisation)
